simulator/modules/decode/decode.cpp

```diff
@@ -99,8 +99,9 @@ void Decode<FuncInstr>::clock( Cycle cycle)
     /* acquiring real information for BPU */
     wp_bp_update->write( instr.get_bp_upd(), cycle);
 
-    bool is_misprediction = instr.is_direct_jump() &&
-    ( !instr.get_bp_data().is_taken || instr.get_bp_data().target != instr.get_decoded_target());
+    bool is_misprediction =
+        ( instr.is_direct_jump() && ( !instr.get_bp_data().is_taken || instr.get_bp_data().target != instr.get_decoded_target()))
+       || instr.is_likely_branch();
```

Comment:
How does it work after misprediction? I feel that it ends up in an infinite loop...
Have you ran any trace?

Reply:
I've run sqrt.s. Then I changed one of the 'bne' instructions to 'bnel' to make sure that it's assumed as 'likely'.

Reply:
Please extract that condition to a new function. Then we'll discuss it.

Reply:
Done. But I don't know if I have chosen a good place for this function or not.